R/utilities.R

```diff
@@ -2646,3 +2646,267 @@ standardize_chr_prefix = function(incoming_vector,
+#' sample, chrom, start, end, and log.ratio. The function can work with either individual or multi-sample seg file. The telomeres are always
+#' excluded from calculation, and centromeres/sex chromosomes can be optionally included or excluded.
+#'
+#' @param this_seg Input data frame of seg file.
+#' @param seg_path Optionally, specify the path to a local seg file.
+#' @param projection Argument specifying the projection of seg file, which will determine chr prefix, chromosome coordinates, and genome size. Default is grch37, but hg38 is also accepted.
+#' @param cutoff The minimum log.ratio for the segment to be considered as CNV. Default is 0.56, which is 1 copy. This value is expected to be positive float of log.ratio for both deletions and amplifications.
+#' @param exclude_sex Boolean argument specifying whether to exclude sex chromosomes from calculation. Default is TRUE.
+#' @param exclude_centromeres Boolean argument specifyng whether to exclude centromeres from calculation. Default is TRUE.
+#'
+#' @return A data frame of sample_id and a float in the range [0..1] indicating the fraction of genome altered by CNV.
+#' @export
+#' @import data.table tidyverse
+#'
+#' @examples
+#' sample_seg = get_sample_cn_segments(this_sample_id = "14-36022T") %>% rename("sample"="ID")
+#' calculate_pga(this_seg = sample_seg)
+#' calculate_pga(this_seg = sample_seg, exclude_sex = FALSE)
+#'
+#' multi_sample_seg = rbind(get_sample_cn_segments(this_sample_id = "14-36022T"),
+#'                          get_sample_cn_segments(this_sample_id = "BLGSP-71-21-00243-01A-11E")) %>%
+#'                          rename("sample"="ID")
+#' GAMBLR::calculate_pga(this_seg = multi_sample_seg)
+#'
+
+calculate_pga = function(this_seg,
+                         seg_path,
+                         projection = "grch37",
+                         cutoff = 0.56,
+                         exclude_sex = TRUE,
+                         exclude_centromeres = TRUE) {
+  # check for required argument
+  if (missing(this_seg) & missing (seg_path)) {
+    stop("Please provide the data frame of seg file or path to the local seg.")
+  }
+
+  # ensure the specified projection is correct and define chromosome coordinates
+  if (projection == "grch37") {
+    chr_coordinates = chromosome_arms_grch37
+  } else if (projection == "hg38") {
+    chr_coordinates = chromosome_arms_hg38
+  } else {
+    stop(
+      "You specified projection that is currently not supported. Please provide seg files in either hg38 or grch37."
+    )
+  }
+
+  # exclude sex chromosomes
+  if (exclude_sex) {
+    chr_coordinates = chr_coordinates %>%
+      dplyr::filter(!grepl("X|Y", chromosome))
+  }
+
+  # does the user's seg file contain centromeres?
+  if (exclude_centromeres) {
+    chr_coordinates = chr_coordinates %>%
+      group_by(chromosome) %>%
+      mutate(start = min(start),
+             end = max(end)) %>%
+      ungroup
+  }
+
+  # total size of genome in this projection
```

Comment:
Suggestion: put this code outside any function such that it runs when GAMBLR is loaded and sets a global variable that can be used from anywhere:

genome_sizes = list(grch37=genome_size_grch37, hg38=genome_size_hg38)

Reply:
This makes sense, thanks! I'll do it in the next update and make this global for GAMBLR